src/lib.rs

```diff
@@ -198,67 +287,200 @@ impl IRCompilerFactory {
+        unsafe {
+            self.funcs
+                .IRCompilerSetHitgroupType(self.me.as_ptr(), hit_group_type)
+        }
+    }
+
+    #[doc(alias = "IRMetalLibSynthesizeIndirectIntersectionFunction")]
+    pub fn synthesize_indirect_intersection_function(&mut self) -> Option<IRMetalLibBinary> {
+        let target_metallib = IRMetalLibBinary::new(self);
+        if unsafe {
+            self.funcs.IRMetalLibSynthesizeIndirectIntersectionFunction(
+                self.me.as_ptr(),
+                target_metallib.me.as_ptr(),
+            )
+        } {
+            Some(target_metallib)
+        } else {
+            None
+        }
+    }
+
+    #[doc(alias = "IRMetalLibSynthesizeIndirectRayDispatchFunction")]
+    pub fn synthesize_indirect_ray_dispatch_function(&mut self) -> Option<IRMetalLibBinary> {
+        let target_metallib = IRMetalLibBinary::new(self);
+        if unsafe {
+            self.funcs.IRMetalLibSynthesizeIndirectRayDispatchFunction(
+                self.me.as_ptr(),
+                target_metallib.me.as_ptr(),
+            )
+        } {
+            Some(target_metallib)
+        } else {
+            None
+        }
+    }
+
+    #[doc(alias = "IRCompilerSetEntryPointName")]
     pub fn set_entry_point_name(&mut self, new_name: &CStr) {
         unsafe {
             self.funcs
-                .IRCompilerSetEntryPointName(self.me, new_name.as_ptr())
+                .IRCompilerSetEntryPointName(self.me.as_ptr(), new_name.as_ptr())
         }
     }
 
+    #[doc(alias = "IRCompilerAllocCompileAndLink")]
     pub fn alloc_compile_and_link(
         &mut self,
         entry_point: &CStr,
         input: &IRObject,
-    ) -> Result<IRObject, Box<dyn std::error::Error>> {
+    ) -> Result<IRObject, CompilerError> {
         let mut error = std::ptr::null_mut();
 
-        let v = unsafe {
+        let v = NonNull::new(unsafe {
             self.funcs.IRCompilerAllocCompileAndLink(
-                self.me,
+                self.me.as_ptr(),
                 entry_point.as_ptr(),
-                input.me,
+                input.me.as_ptr(),
                 &mut error,
             )
-        };
+        });
 
         if error.is_null() {
             Ok(IRObject {
-                me: v,
+                me: v.unwrap(),
                 funcs: input.funcs.clone(),
             })
         } else {
-            panic!("{:?}", error);
+            let error = IRError::from_ptr(error, self.funcs.clone());
+            Err(CompilerError(error.code()))
+        }
+    }
+}
+
+struct IRError {
+    me: NonNull<bindings::IRError>,
+    funcs: Arc<bindings::metal_irconverter>,
+}
+
+impl Drop for IRError {
+    #[doc(alias = "IRErrorDestroy")]
+    fn drop(&mut self) {
+        unsafe { self.funcs.IRErrorDestroy(self.me.as_ptr()) };
+    }
+}
+
+impl IRError {
+    pub fn from_ptr(ptr: *mut bindings::IRError, funcs: Arc<bindings::metal_irconverter>) -> Self {
```

Comment:
Fwiw we could also make this take a `NonNull`. Then the above `error.is_null()` becomes a cleaner:

```rust
if let Some(error) = NonNull::new(error) {
    Err(IRError::from_ptr(error, ...))
} else {
   Ok(...)
}
```

Reply:
Doesn't seem to make much of a difference in how the code looks, but working with a NonNull instead of a raw pointer might be good